components/TextFields/src/MDCTextInputBehavior.h

```diff
@@ -0,0 +1,163 @@
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#import <UIKit/UIKit.h>
+
+@protocol MDCTextInput;
+@protocol MDCTextInputCharacterCounter;
+
+/**
+ Presentation styles for a text input. The style determines specific aspects of the text
+ input, such as sizing, placeholder placement and behavior, layout, etc.
+ */
+typedef NS_ENUM(NSUInteger, MDCTextInputPresentationStyle) {
+  /**
+   Default style with an inline placeholder (that disappears when text is entered) and character
+   count / limit below text.
+   */
+  MDCTextInputPresentationStyleDefault = 0,
+
+  /**
+   The placeholder text is laid out inline but will float above the field when there is content or
+   the field is being edited. The character count is below text. The Material Design guidelines
+   call this 'Floating inline labels.'
+   https://material.io/guidelines/components/text-fields.html#text-fields-labels
+   */
+  MDCTextInputPresentationStyleFloatingPlaceholder,
+
+  /** The placeholder is laid out inline and the character count is also inline to the right. */
+  MDCTextInputPresentationStyleFullWidth,
+};
+
+/**
+ Material Design themed states for textInputs. The logic for 'automagic' error states changes:
+ underline color, underline text color.
+ https://www.google.com/design/spec/components/text-fields.html#text-fields-single-line-text-field
+ */
+@interface MDCTextInputBehavior : NSObject <NSCoding, NSCopying>
+
+/** The text input the behavior is managing. */
+@property(nonatomic, nullable, weak) UIView<MDCTextInput> *textInput NS_SWIFT_NAME(input);
+
+/**
+ The color used to denote error state in the underline, the errorText's label, the plceholder and
+ the character count label.
+
+ Default is red.
+ */
+@property(nonatomic, nullable, strong) UIColor *errorColor UI_APPEARANCE_SELECTOR;
+
+/**
+ Controls when the underline will be shown.
+
+ The underline is an overlay that can be hidden depending on the editing state of the input text.
+
+ Default is UITextFieldViewModeAlways.
+ */
+@property(nonatomic, assign) UITextFieldViewMode underlineViewMode NS_SWIFT_NAME(underlineMode)
+    UI_APPEARANCE_SELECTOR;
+
+/**
+ Optional, custom character counter.
+
+ Instead of relying on the default character count which is naive (counts each character regardless
+ of context), this object can instead choose to do sophisticated counting (ie: ignoring whitespace,
+ ignoring url strings, etc).
+ */
+@property(nonatomic, nullable, weak) IBInspectable id<MDCTextInputCharacterCounter>
```

Comment:
I mean instead of doing some internal logic when this is nil just put that internal logic in a MDCTextInputCharacterCounter conforming object and set that as the default.